File: src/main/java/miragefairy2019/mod/modules/fairy/ItemFairy.java
package miragefairy2019.mod.modules.fairy;

import miragefairy2019.mod.api.fairy.IFairyType;
import miragefairy2019.mod.api.fairy.IItemFairy;
import miragefairy2019.mod.api.fairyweapon.item.IItemFairyWeapon;
import miragefairy2019.mod.lib.UtilsMinecraft;
import miragefairy2019.mod.lib.multi.ItemMulti;
import miragefairy2019.modkt.api.IManaType;
import miragefairy2019.modkt.impl.ManaSetKt;
import miragefairy2019.modkt.impl.ManaType;
import miragefairy2019.modkt.impl.ManaTypeKt;
import miragefairy2019.modkt.impl.fairy.AbilityTypeKt;
import mirrg.boron.util.UtilsString;
import net.minecraft.client.Minecraft;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumHand;
import net.minecraft.util.text.Style;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.util.text.TextFormatting;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.List;
import java.util.Optional;

import static net.minecraft.util.text.TextFormatting.*;

public class ItemFairy extends ItemMulti<VariantFairy> implements IItemFairy {

    @Override
    public Optional<IFairyType> getMirageFairy2019Fairy(ItemStack itemStack) {
        VariantFairy variant = getVariant(itemStack).orElse(null);
        if (variant == null) return Optional.empty();
        return Optional.of(variant.type);
    }

    @Override
    public String getItemStackDisplayName(ItemStack itemStack) {
        VariantFairy variant = getVariant(itemStack).orElse(null);
        if (variant == null) return UtilsMinecraft.translateToLocal(getUnlocalizedName() + ".name");
        return UtilsMinecraft.translateToLocalFormatted(getUnlocalizedName() + ".format", variant.type.getDisplayName().getFormattedText());
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void addInformation(ItemStack itemStack, World world, List<String> tooltip, ITooltipFlag flag) {
        VariantFairy variant = getVariant(itemStack).orElse(null);
        if (variant == null) return;

        if (!flag.isAdvanced()) {
            tooltip.add("" + GREEN + "No: " + variant.type.id + " (" + variant.type.modid + ")");
            tooltip.add("" + "Type:"
                    + " " + GOLD + UtilsString.repeat("★", variant.type.rare) + getRankColor(variant) + UtilsString.repeat("★", variant.type.rank - 1)
                    + " " + WHITE + variant.type.name);
        } else {
            tooltip.add("" + GREEN + "No: " + variant.type.id + " (" + variant.type.modid + ")");
            tooltip.add("" + "Type:"
                    + " " + GOLD + UtilsString.repeat("★", variant.type.rare) + getRankColor(variant) + UtilsString.repeat("★", variant.type.rank - 1)
                    + " " + GOLD + "Rare." + variant.type.rare
                    + " " + getRankColor(variant) + "Rank." + variant.type.rank
                    + " " + WHITE + variant.type.name);
        }

        if (!flag.isAdvanced()) {
            tooltip.add("    " + f1(ManaType.Companion.getShine(), variant));
            tooltip.add("" + f1(ManaType.Companion.getFire(), variant) + "    " + f1(ManaType.Companion.getWind(), variant));
            tooltip.add("" + f1(ManaType.Companion.getGaia(), variant) + "    " + f1(ManaType.Companion.getAqua(), variant));
            tooltip.add("    " + f1(ManaType.Companion.getDark(), variant));
        } else {
            tooltip.add("        " + f2(ManaType.Companion.getShine(), variant));
            tooltip.add("" + f2(ManaType.Companion.getFire(), variant) + "    " + f2(ManaType.Companion.getWind(), variant));
            tooltip.add("" + f2(ManaType.Companion.getGaia(), variant) + "    " + f2(ManaType.Companion.getAqua(), variant));
            tooltip.add("        " + f2(ManaType.Companion.getDark(), variant));
        }

        if (!flag.isAdvanced()) {
            tooltip.add("" + DARK_PURPLE + "Cost: " + String.format("%.1f", variant.type.cost));
        } else {
            tooltip.add("" + DARK_PURPLE + "Cost: " + String.format("%.3f", variant.type.cost));
        }

        if (!flag.isAdvanced()) {
            tooltip.add(new TextComponentString("")
                    .setStyle(new Style().setColor(GREEN))
                    .appendText("Abilities: ")
                    .appendSibling(variant.type.abilitySet.getAbilities()
                            .filter(tuple -> format(tuple.y) >= 10)
                            .sorted((a, b) -> -a.y.compareTo(b.y))
                            .map(t -> {
                                return new TextComponentString("")
                                        .appendSibling(AbilityTypeKt.getDisplayName(t.x))
                                        .appendText(String.format("(%s)", format(t.y)));
                            })
                            .sandwich(new TextComponentString(", "))
                            .apply(tcs -> {
                                TextComponentString textComponent = new TextComponentString("");
                                tcs.forEach(textComponent::appendSibling);
                                return textComponent;
                            }))
                    .getFormattedText());
        } else {
            tooltip.add(new TextComponentString("")
                    .setStyle(new Style().setColor(GREEN))
                    .appendText("Abilities: ")
                    .appendSibling(variant.type.abilitySet.getAbilities()
                            .filter(tuple -> format(tuple.y) >= 10)
                            .sorted((a, b) -> -a.y.compareTo(b.y))
                            .map(t -> {
                                return new TextComponentString("")
                                        .appendSibling(AbilityTypeKt.getDisplayName(t.x))
                                        .appendText(String.format("(%.3f)", t.y));
                            })
                            .sandwich(new TextComponentString(", "))
                            .apply(tcs -> {
                                TextComponentString textComponent = new TextComponentString("");
                                tcs.forEach(textComponent::appendSibling);
                                return textComponent;
                            }))
                    .getFormattedText());
        }

        // 妖精武器のステータス
        EntityPlayer player = Minecraft.getMinecraft().player;
        if (player != null) {
            ItemStack itemStackFairyWeapon;
            itemStackFairyWeapon = player.getHeldItem(EnumHand.MAIN_HAND);
            if (itemStackFairyWeapon.getItem() instanceof IItemFairyWeapon) {
                tooltip.add(new TextComponentString("Magic: ")
                        .setStyle(new Style().setColor(BLUE))
                        .appendSibling(((IItemFairyWeapon) itemStackFairyWeapon.getItem()).getFairyMagicDisplayName(itemStackFairyWeapon)
                                .setStyle(new Style().setColor(AQUA).setBold(true)))
                        .appendText(" of ")
                        .appendSibling(new TextComponentString(itemStackFairyWeapon.getDisplayName())
                                .setStyle(new Style().setColor(WHITE)))
                        .getFormattedText());
                ((IItemFairyWeapon) itemStackFairyWeapon.getItem()).addInformationFairyWeapon(itemStackFairyWeapon, itemStack, variant.type, world, tooltip, flag);
            }
            itemStackFairyWeapon = player.getHeldItem(EnumHand.OFF_HAND);
            if (itemStackFairyWeapon.getItem() instanceof IItemFairyWeapon) {
                tooltip.add(new TextComponentString("Magic: ")
                        .setStyle(new Style().setColor(BLUE))
                        .appendSibling(((IItemFairyWeapon) itemStackFairyWeapon.getItem()).getFairyMagicDisplayName(itemStackFairyWeapon)
                                .setStyle(new Style().setColor(AQUA).setBold(true)))
                        .appendText(" of ")
                        .appendSibling(new TextComponentString(itemStackFairyWeapon.getDisplayName())
                                .setStyle(new Style().setColor(WHITE)))
                        .getFormattedText());
                ((IItemFairyWeapon) itemStackFairyWeapon.getItem()).addInformationFairyWeapon(itemStackFairyWeapon, itemStack, variant.type, world, tooltip, flag);
            }
        }

    }

    private TextFormatting getRankColor(VariantFairy variant) {
        switch (variant.type.rank) {
            case 1:
                return RED;
            case 2:
                return BLUE;
            case 3:
                return GREEN;
            case 4:
                return YELLOW;
            case 5:
                return DARK_GRAY;
            case 6:
                return WHITE;
            case 7:
                return AQUA;
            default:
                return LIGHT_PURPLE;
        }
    }

    private int format(double value) {
        int value2 = (int) value;
        if (value2 == 0 && value > 0) value2 = 1;
        return value2;
    }

    private String f1(IManaType manaType, VariantFairy variant) {
        return "" + manaType.getTextColor() + String.format("%4d", format(ManaSetKt.getMana(variant.type.manaSet, manaType)));
    }

    private String f2(IManaType manaType, VariantFairy variant) {
        return "" + manaType.getTextColor() + ManaTypeKt.getDisplayName(manaType).getUnformattedText() + ":" + String.format("%.3f", ManaSetKt.getMana(variant.type.manaSet, manaType));
    }

}

File: src/main/java/miragefairy2019/modkt/modules/fairy/ItemFairy.kt
package miragefairy2019.modkt.modules.fairy

import miragefairy2019.libkt.TextComponentBuilder
import miragefairy2019.libkt.bold
import miragefairy2019.libkt.buildText
import miragefairy2019.libkt.color
import miragefairy2019.mod.api.fairy.IItemFairy
import miragefairy2019.mod.api.fairyweapon.item.IItemFairyWeapon
import miragefairy2019.mod.lib.UtilsMinecraft
import miragefairy2019.mod.lib.multi.ItemMulti
import miragefairy2019.modkt.api.fairy.IFairyType
import miragefairy2019.modkt.api.mana.IManaType
import miragefairy2019.modkt.api.mana.ManaTypes
import miragefairy2019.modkt.impl.fairy.displayName
import miragefairy2019.modkt.impl.getMana
import miragefairy2019.modkt.impl.mana.displayName
import mirrg.boron.util.UtilsString
import net.minecraft.client.Minecraft
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumHand
import net.minecraft.util.text.TextComponentString
import net.minecraft.util.text.TextFormatting.*
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import java.util.*

class ItemFairy : ItemMulti<VariantFairy>(), IItemFairy {
    override fun getMirageFairy2019Fairy(itemStack: ItemStack): Optional<IFairyType> = getVariant(itemStack).map { it.type }
    override fun getItemStackDisplayName(itemStack: ItemStack): String = getMirageFairy2019Fairy(itemStack).map { UtilsMinecraft.translateToLocalFormatted("$unlocalizedName.format", it.displayName.formattedText) }.orElseGet { UtilsMinecraft.translateToLocal("$unlocalizedName.name") }

    @SideOnly(Side.CLIENT)
    override fun addInformation(itemStack: ItemStack, world: World?, tooltip: MutableList<String>, flag: ITooltipFlag) {
        val variant = getVariant(itemStack).orElse(null) ?: return
        fun tooltip(block: TextComponentBuilder.() -> Unit) = run { tooltip.add(buildText { block() }.formattedText); Unit }
        fun formatInt(value: Double) = value.toInt().let { if (it == 0 && value > 0) 1 else it }


        // 番号　MOD名
        if (flag.isAdvanced) {
            tooltip {
                text("No: ${variant.id} (${variant.type.breed?.resourceDomain ?: "unknown"})").color(GREEN)
            }
        }

        // レア　ランク　品種名
        tooltip {
            fun getRankColor(variant: VariantFairy) = when (variant.rank) {
                1 -> RED
                2 -> BLUE
                3 -> GREEN
                4 -> YELLOW
                5 -> DARK_GRAY
                6 -> WHITE
                7 -> AQUA
                else -> LIGHT_PURPLE
            }

            text("Type: ")
            text(UtilsString.repeat("★", variant.rare)).color(GOLD)
            if (flag.isAdvanced) {
                text(UtilsString.repeat("★", variant.rank - 1)).color(getRankColor(variant))
            } else {
                text(UtilsString.repeat("★", variant.rank - 1)).color(GOLD)
            }
            if (flag.isAdvanced) text(" Rare.${variant.rare}").color(GOLD)
            if (flag.isAdvanced) text(" Rank.${variant.rank}").color(getRankColor(variant))
            text(" ${variant.type.breed?.resourcePath ?: "unknown"}").color(WHITE)
        }

        // マナ
        tooltip {
            text("Mana: ").color(AQUA)
        }
        if (flag.isAdvanced) {
            fun f(manaType: IManaType) = buildText { format("%s:%.3f", manaType.displayName.unformattedText, variant.type.manaSet.getMana(manaType)).color(manaType.textColor) }
            tooltip {
                text("        ")
                text(f(ManaTypes.shine))
            }
            tooltip {
                text(f(ManaTypes.fire))
                text("    ")
                text(f(ManaTypes.wind))
            }
            tooltip {
                text(f(ManaTypes.gaia))
                text("    ")
                text(f(ManaTypes.aqua))
            }
            tooltip {
                text("        ")
                text(f(ManaTypes.dark))
            }
        } else {
            fun f(manaType: IManaType) = buildText { format("%4d", formatInt(variant.type.manaSet.getMana(manaType))).color(manaType.textColor) }
            tooltip {
                text("    ")
                text(f(ManaTypes.shine))
            }
            tooltip {
                text(f(ManaTypes.fire))
                text("    ")
                text(f(ManaTypes.wind))
            }
            tooltip {
                text(f(ManaTypes.gaia))
                text("    ")
                text(f(ManaTypes.aqua))
            }
            tooltip {
                text("    ")
                text(f(ManaTypes.dark))
            }
        }

        // コスト
        tooltip {
            if (flag.isAdvanced) {
                format("Cost: %.3f", variant.type.cost).color(DARK_PURPLE)
            } else {
                format("Cost: %.0f", variant.type.cost).color(DARK_PURPLE)
            }
        }

        // エルグ
        fun <T> Iterable<T>.sandwich(separator: T) = mapIndexed { i, it -> if (i == 0) listOf(it) else listOf(separator, it) }.flatten()
        tooltip {
            text {
                text("Erg: ")
                variant.type.ergSet.entries
                        .filter { formatInt(it.power) >= 10 }
                        .sortedByDescending { it.power }
                        .map {
                            buildText {
                                text(it.type.displayName)
                                if (flag.isAdvanced) format("(%.3f)", it.power)
                            }
                        }
                        .sandwich(buildText { text(", ") })
                        .forEach { text(it) }
            }.color(GREEN)
        }

        // 妖精武器のステータス
        Minecraft.getMinecraft().player?.let { player ->
            fun f(itemStackFairyWeapon: ItemStack) {
                val item = itemStackFairyWeapon.item as? IItemFairyWeapon ?: return
                tooltip {
                    text {
                        text("Magic: ")
                        text(item.getFairyMagicDisplayName(itemStackFairyWeapon)).color(AQUA).bold()
                        text(" of ")
                        text(TextComponentString(itemStackFairyWeapon.displayName)).color(WHITE)
                    }.color(BLUE)
                }
                item.addInformationFairyWeapon(itemStackFairyWeapon, itemStack, variant.type, world, tooltip, flag)
            }
            f(player.getHeldItem(EnumHand.MAIN_HAND))
            f(player.getHeldItem(EnumHand.OFF_HAND))
        }

    }
}
